dotnet/src/SemanticKernel/Reliability/DefaultHttpRetryHandler.cs

```diff
@@ -0,0 +1,259 @@
+    /// </summary>
+    /// <param name="originalRequest">The previous <see cref="HttpRequestMessage"/> needs to be copy.</param>
+    /// <returns>The <see cref="HttpRequestMessage"/>.</returns>
+    /// <remarks>
+    /// Re-issue a new HTTP request with the previous request's headers and properties
+    /// </remarks>
+    internal static async Task<HttpRequestMessage> CloneAsync(HttpRequestMessage originalRequest)
+    {
+        var newRequest = new HttpRequestMessage(originalRequest.Method, originalRequest.RequestUri);
+
+        // Copy request headers.
+        foreach (var header in originalRequest.Headers)
+        {
+            newRequest.Headers.TryAddWithoutValidation(header.Key, header.Value);
+        }
+
+        // Copy request properties.
+        foreach (var property in originalRequest.Properties)
+        {
+            newRequest.Properties.Add(property);
+        }
+
+        // Set Content if previous request had one.
+        if (originalRequest.Content != null)
+        {
+            // HttpClient doesn't rewind streams and we have to explicitly do so.
+            await originalRequest.Content.ReadAsStreamAsync().ContinueWith(t =>
+            {
+                if (t.Result.CanSeek) { t.Result.Seek(0, SeekOrigin.Begin); }
+
+                newRequest.Content = new StreamContent(t.Result);
+            }, TaskScheduler.Current).ConfigureAwait(false);
+
+            // Copy content headers.
+            if (originalRequest.Content.Headers != null)
+            {
+                foreach (var contentHeader in originalRequest.Content.Headers)
+                {
+                    newRequest.Content.Headers.TryAddWithoutValidation(contentHeader.Key, contentHeader.Value);
+                }
+            }
+        }
+
+        return newRequest;
+    }
+
+    /// <summary>
+    /// Executes the action with retry logic
+    /// </summary>
+    /// <remarks>
+    /// The request is retried if it throws an exception that is a retryable exception.
+    /// If the request throws an exception that is not a retryable exception, it is not retried.
+    /// If the request returns a response with a retryable error code, it is retried.
+    /// If the request returns a response with a non-retryable error code, it is not retried.
+    /// If the exception contains a RetryAfter header, the request is retried after the specified delay.
+    /// If configured to use exponential backoff, the delay is doubled for each retry.
+    /// </remarks>
+    /// <param name="request">The request.</param>
+    /// <param name="cancellationToken">The cancellation token.</param>
+    protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+    {
+        int retryCount = 0;
+
+        var start = this._timeProvider.GetCurrentTime();
+        while (true)
```

Comment:
I'd rather keep the logic contained to the loop to avoid having to delay before making the request and complicating the check for if enough time is allowed. Logically, the first attempt (does not count as a retry) always happens so it makes sense to me to present the code in such a way. Especially since we need to determine whether we return the response as-is or clone before waiting.

Reply:
Resolving as-is.